docs/decisions/0040-OTel-semantic-convention.md

```diff
@@ -0,0 +1,628 @@
+
+### Where do we create the activities
+
+#### Connector/AI Service
+
+AI services employ common interfaces and typically feature straightforward implementations, thereby resulting in lower maintenance overhead. Nonetheless, acquiring comprehensive model-related data and complete LLM requests/responses can be challenging within the services.
+
+#### Client
+
+Client implementations may exhibit variations; however, they generally encompass a similar set of methods for text and chat completions. Furthermore, obtaining model data, such as model name and provider, as well as response metadata, including token usage, is more convenient when performed within the clients.
+
+#### Comparison
+
+Clients represent the final layer in the stack through which all kernel operations reach the models. As such, it would be more intuitive to incorporate instrumentation for all LLM requests at the client level. Furthermore, some clients already generate metrics. However, due to the lack of shared interfaces among clients, there will be additional implementation and maintenance overhead.
+
+In contrast, connectors (or AI services) share common interfaces, specifically the `IChatCompletionService` and `ITextGenerationService`. Nevertheless, AI services rely on their underlying clients to execute LLM requests, which could potentially result in inconsistencies where telemetry data originates from different layers in the stack if a client is instrumented (e.g., Azure OpenAI vs. `HuggingFaceTextGenerationService`).
+
+|      | Connectors/Services                                                                                           | Clients                                                                                                          |
+| ---- | ------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------- |
+| Pros | 1. Uniform implementation across all services.                                                                | 1. Enhanced intuitiveness regarding the overall data flow.<br>2. Complete model specs and LLM requests/response. |
+| Cons | 1. Inconsistencies in the origins of telemetry data.<br>2. Incomplete model specs and LLM requests/responses. | 1. Separate implementations may lead to increased maintenance overhead.                                          |
+
+## Decision Outcome
+
+TBD
+
+## Appendix
+
+### `PromptExecutionSettings`
+
+```C#
+public class PromptExecutionSettings
+{
+    ...
+    [JsonPropertyName("is_event_collection_enable")]
+    public bool IsEventCollectionEnabled
```

Comment:
this is a great start, but I think we'll need to have more options:

- don't collect events at all
- collect events, but without sensitive parts. E.g. we might collect `{"role":"assistant","tool_calls":[{"id":"call_hHM72v9f1JprJBStycQC4Svz","function":{"name":"get_weather","arguments":"{\"location\": \"REFDACTED\"}"},"type":"function"}]}` or `{"role":"user", "content":"REDACTED"}`

so maybe enum would work better, like

```csharp
enum EventCollectionLevel 
{
   None,
   RedactSensitiveData,
   All,
}
```

Reply:
I think we can explore this level of granularity in the future. For now, I have provided the option to opt in to sensitive data collection instead of event collection.